File: .NET/BinToBloom Backend/Controllers/TrackingController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using BinToBloom_Backend.Data;
using BinToBloom_Backend.Models.Entities;
using System.Security.Claims;

namespace BinToBloom_Backend.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class TrackingController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public TrackingController(ApplicationDbContext context)
        {
            _context = context;
        }

        [HttpGet("{pickupId}")]
        [Authorize]
        public async Task<IActionResult> GetTrackingInfo(int pickupId)
        {
            try
            {
                var userIdStr = User.FindFirstValue(ClaimTypes.NameIdentifier);
                if (string.IsNullOrEmpty(userIdStr)) return Unauthorized("User not authenticated");
                int currentUserId = int.Parse(userIdStr);

                var pickup = await _context.PickupRequests
                    .Include(p => p.User)
                    .Include(p => p.Collector)
                    .ThenInclude(c => c.User)
                    .FirstOrDefaultAsync(p => p.PickupId == pickupId);

                if (pickup == null)
                {
                    return BadRequest(new { message = "Pickup not found" });
                }

                // Check permissions: Only the user who requested OR the assigned collector can track
                // If the user is an ADMIN, they might also see it, but primary logic is User/Collector
                bool isAuthorized = pickup.UserId == currentUserId || 
                                   (pickup.Collector != null && pickup.Collector.UserId == currentUserId);

                if (!isAuthorized && !User.IsInRole("ADMIN"))
                {
                    return StatusCode(403, new { message = "Access denied" });
                }

                var trackingInfo = new Dictionary<string, object>
                {
                    { "pickupId", pickup.PickupId },
                    { "status", pickup.PickupStatus },
                    { "pickupLocation", new { lat = pickup.Latitude ?? 0, lng = pickup.Longitude ?? 0 } }
                };

                if (pickup.Collector != null)
                {
                    trackingInfo.Add("collectorLocation", new { lat = pickup.Collector.CurrentLat ?? 0, lng = pickup.Collector.CurrentLng ?? 0 });
                    trackingInfo.Add("collectorName", pickup.Collector.User.Name);
                    trackingInfo.Add("collectorPhone", pickup.Collector.User.Phone);
                }

                return Ok(trackingInfo);
            }
            catch (Exception e)
            {
                return BadRequest(new { message = "Failed to get tracking info: " + e.Message });
            }
        }
    }
}

File: Springboot/bintobloom-backend/src/main/java/com/bintobloom/controller/TrackingController.java
package com.bintobloom.controller;

import com.bintobloom.entity.PickupRequest;
import com.bintobloom.entity.User;
import com.bintobloom.repository.PickupRequestRepository;
import com.bintobloom.repository.UserRepository;
import com.bintobloom.security.UserPrincipal;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/tracking")
@CrossOrigin(origins = "*", maxAge = 3600)
public class TrackingController {

    @Autowired
    private PickupRequestRepository pickupRequestRepository;

    @Autowired
    private UserRepository userRepository;

    @GetMapping("/{pickupId}")
    public ResponseEntity<?> getTrackingInfo(@PathVariable Long pickupId,
            @AuthenticationPrincipal UserPrincipal userPrincipal) {
        try {
            Optional<PickupRequest> pickupOptional = pickupRequestRepository.findById(pickupId);
            if (pickupOptional.isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of("message", "Pickup not found"));
            }

            PickupRequest pickup = pickupOptional.get();

            // Only the user who requested or the collector assigned can track
            if (!pickup.getUser().getUserId().equals(userPrincipal.getId()) &&
                    (pickup.getCollector() == null
                            || !pickup.getCollector().getUserId().equals(userPrincipal.getId()))) {
                return ResponseEntity.status(403).body(Map.of("message", "Access denied"));
            }

            Map<String, Object> trackingInfo = new HashMap<>();
            trackingInfo.put("pickupId", pickup.getPickupId());
            trackingInfo.put("status", pickup.getPickupStatus());
            trackingInfo.put("pickupLocation", Map.of(
                    "lat", pickup.getLatitude() != null ? pickup.getLatitude() : 0,
                    "lng", pickup.getLongitude() != null ? pickup.getLongitude() : 0));

            if (pickup.getCollector() != null) {
                User collector = pickup.getCollector();
                trackingInfo.put("collectorLocation", Map.of(
                        "lat", collector.getLatitude() != null ? collector.getLatitude() : 0,
                        "lng", collector.getLongitude() != null ? collector.getLongitude() : 0));
                trackingInfo.put("collectorName", collector.getName());
                trackingInfo.put("collectorPhone", collector.getPhone());
            }

            return ResponseEntity.ok(trackingInfo);
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Failed to get tracking info: " + e.getMessage()));
        }
    }
}
